x-pack/legacy/plugins/siem/server/lib/detection_engine/alerts/utils.ts

```diff
@@ -56,7 +60,9 @@ export const singleBulkIndex = async (
     {
       index: {
         _index: process.env.SIGNALS_INDEX || DEFAULT_SIGNALS_INDEX,
-        _id: doc._id,
+        _id: createHash('sha256')
+          .update(id.concat(doc._id))
```

Comment:
Is `id` the `rule.id` which is an autogenerated GUID from the alerting system?

If so, if the user deletes the rule altogether and re-creates it exactly as is, wouldn't that create a different `id` and then you will end up with duplicates suddenly?

I would try to use everything within each document so the document its self is the thing that has a unique but stable reference in which external parts of the system cannot cause duplicates.

Reply:
Yes that would lead to duplicates so not what we were looking for. And I now think I'm confused as to the original problem statement.

Using the signals source event `_id` would prevent duplicates from showing up in the signals index, but would drop signals with overlapping source events, so that much makes sense as to why we wouldn't reuse the source event _id.

I think you mentioned previously how maybe generating the signal id as a function of the signal documents timestamp at creation plus the source _id would be a better way?

Reply:
I guess that would lead to duplicates too

Reply:
I read in one of the blogposts about Logstash utilizing some fingerprinting filter to identify the original event.. This would be a combination of fields from the source document, but I feel that might also lead to signals with overlapping source events being dropped which wouldn't be ideal.

Reply:
If this is true:

The _id might not be unique among indexes.
If it's not unique then we end up with records overwriting other records when they go from N indexes down to 1 signals index.

Then we want to apply a technique for de-duplication that is fast for large amounts of signals that works in an idempotent manner where you can re-run the same signal jobs and the same output in the index occurs.

Reply:
Updating for historical purposes - incorporating rule_id (which is different from the `id` of the rule) in order to add the ability for the same source event to be triggered by different rules, and thus generate different signals. Without this option, rules could be triggered by the same source event, but only one signal (the first rule triggered by the event) will appear in the signals index (unless the source events document version was updated). With the addition of rule_id we can now create signals based on different rules that incorporate the same source event.